### PR TITLE
Ingm 404 make new example about ether cat co e connection

### PR DESCRIPTION
**Pull request by WoodElv**
### Description

A new script is added into the examples folder for a drive connection using EtherCAT - CoE.
Also, unit tests are added for this script.

### Type of change

- [x] Script for a drive connection using EtherCAT - CoE.

### Tests
- [x] Unit test if the CoE connection successes.
- [x] Unit test if the CoE connection failed.
- [x] Unit test if there is a ConnectionError exception.

### Documentation

Please update the documentation.

- [x] Use [type hints](https://docs.python.org/3/library/typing.html) for every function and argument.
- [x] Build documentation locally to verify changes.

### Code formatting

- [x] Use black package to format the code: `black -l 100 ingeniamotion tests`. It is recommended to configure the code editor to automatically format the code using black with a max length line of 100.

### CODE REVIEW

File: examples/connect_ecat_coe.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+    try:
+        slave_id_list = mc.communication.scan_servos_ethercat(interface_selected)
+    except ConnectionError as e:
+        print(e)
+        return
+
+    if not slave_id_list:
+        print(
+            f"Any slave is detected using the interface: {interface_list_human_format[ecat_coe_conf['interface_index']]}."
```
Comment (MartinALL91) [inline examples/connect_ecat_coe.py:30]: No slave detected on interface

File: examples/connect_ecat_coe.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        print(f"Found slaves: {slave_id_list}")
+
+    try:
+        mc.communication.connect_servo_ethercat(
+            interface_selected, ecat_coe_conf["slave_id"], ecat_coe_conf["dictionary"]
+        )
+    except FileNotFoundError as e:
+        print(e)
+        return
+    if "default" not in mc.servos:
```
Comment (MartinALL91) [inline examples/connect_ecat_coe.py:43]: Something has had to go wrong to reach this point. I would expand the try/except block.

File: examples/connect_ecat_coe.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        print("Drive is not connected.")
+        return
+    else:
+        print("Drive is connected.")
+
+    mc.communication.disconnect()
+    if "default" in mc.servos:
```
Comment (MartinALL91) [inline examples/connect_ecat_coe.py:50]: In what case could this fail?

File: examples/connect_ecat_coe.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        print("Drive disconnection failed.")
+    else:
+        print("The drive has been disconnected.")
+
+
+if __name__ == "__main__":
+    # Replace the ecat_coe_conf parameters
+    ecat_coe_conf = {
+        "interface_index": 3,
```
Comment (mjedieckmann) [inline examples/connect_ecat_coe.py:59]: Can we hardcode the index like that? Aren't the available interfaces and their indices dependent on the machine you run the code on?

**Review by mjedieckmann — APPROVED** [review]

File: examples/connect_ecat_coe.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    print(f"- Index interface: {ecat_coe_conf['interface_index']}")
+    print(f"- Real name: {interface_selected}")
+    print(
+        f"- Human-readable format name: {interface_list_human_format[ecat_coe_conf['interface_index']]}"
+    )
+
+    try:
+        slave_id_list = mc.communication.scan_servos_ethercat(interface_selected)
+    except ConnectionError as e:
+        print(e)
```
Comment (GullinBustin) [inline examples/connect_ecat_coe.py:25]: Why do you use a try-except instead of do nothing and allow raise an exception if something is wrong? It is not a good practice.

**Review by GullinBustin — CHANGES_REQUESTED** [review]

File: examples/connect_ecat_coe.py (hunk 1 above)
Comment (GullinBustin) [reply examples/connect_ecat_coe.py:30]: Did you think about use logger or ingenialogger instead prints?

Comment (WoodElv) [reply examples/connect_ecat_coe.py:30]: Yes, but there are another loggers that come from the library, and for an example script I think they make noise. I guess that's the reason that there is any logger in any other example that is already made.

File: examples/connect_ecat_coe.py (hunk 5 above)
Comment (WoodElv) [reply examples/connect_ecat_coe.py:25]: Done

File: examples/connect_ecat_coe.py (hunk 3 above)
Comment (WoodElv) [reply examples/connect_ecat_coe.py:50]: Done

File: examples/connect_ecat_coe.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
+from ingeniamotion import MotionController
+
+
+def connect_ethercat_coe(interface_index: int, slave_id: int, dictionary_path: str) -> None:
+    mc = MotionController()
+
+    interface_list_human_format = mc.communication.get_interface_name_list()
+    print("List of interfaces - Human-readable format:")
```
Comment (pzinemanas) [inline examples/connect_ecat_coe.py:8]: I'd remove the "Human-readable format"

File: examples/connect_ecat_coe.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    for index, interface in enumerate(interface_list_human_format):
+        print(f"{index}: {interface}")
+
+    interface_selected = mc.communication.get_ifname_by_index(interface_index)
+    print("Interface selected:")
+    print(f"- Index interface: {interface_index}")
+    print(f"- Real name: {interface_selected}")
```
Comment (pzinemanas) [inline examples/connect_ecat_coe.py:15]: Interface identifier?

File: examples/connect_ecat_coe.py (hunk 6 above)
Comment (GullinBustin) [reply examples/connect_ecat_coe.py:8]: It is an example, not a script. Add prints or logs could be good, but sometimes adding comments in the code is better than adding prints. No changes are required, but I think it is good to remember for future examples.

**Review by GullinBustin — APPROVED** [review]

File: examples/connect_ecat_coe.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
+from ingeniamotion import MotionController
+
+
+def connect_ethercat_coe(interface_index: int, slave_id: int, dictionary_path: str) -> None:
+    mc = MotionController()
+
+    interface_list_human_format = mc.communication.get_interface_name_list()
```
[the remark this answers is not in the document]
Comment (WoodElv) [reply examples/connect_ecat_coe.py:7]: Done

File: examples/connect_ecat_coe.py (hunk 7 above)
Comment (WoodElv) [reply examples/connect_ecat_coe.py:15]: Done

File: examples/connect_ecat_coe.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    print("List of interfaces - Human-readable format:")
+    for index, interface in enumerate(interface_list_human_format):
+        print(f"{index}: {interface}")
+
+    interface_selected = mc.communication.get_ifname_by_index(interface_index)
+    print("Interface selected:")
+    print(f"- Index interface: {interface_index}")
+    print(f"- Real name: {interface_selected}")
+    print(f"- Human-readable format name: {interface_list_human_format[interface_index]}")
```
[the remark this answers is not in the document]
Comment (WoodElv) [reply examples/connect_ecat_coe.py:16]: Done

**Review by MartinALL91 — APPROVED** [review]

**Review by pzinemanas — APPROVED** [review]

File: examples/connect_ecat_coe.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        print(f"No slave detected on interface: {interface_list[interface_index]}")
+        return
+    else:
+        print(f"Found slaves: {slave_id_list}")
+
+    try:
+        mc.communication.connect_servo_ethercat(interface_selected, slave_id, dictionary_path)
+    except FileNotFoundError as e:
+        print(e)
+        return
```
Comment (GullinBustin) [inline examples/connect_ecat_coe.py:30]: Why did you add a try-except? I think it's better to raise the exception. The behavior is very similar, but now seems that the file not found is an expected behavior.

**Review by GullinBustin — CHANGES_REQUESTED** [review]

**Review by GullinBustin — APPROVED** [review]
